Mathlib/LinearAlgebra/RootSystem/RootPairingCat.lean

```diff
@@ -0,0 +1,144 @@
+/-
+Copyright (c) 2024 Scott Carnahan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Carnahan
+-/
+import Mathlib.LinearAlgebra.RootSystem.Defs
+import Mathlib.CategoryTheory.Category.Basic
+
+/-!
+# The category of root pairings
+This file defines the category of root pairings, following the definition of category of root data
+given in SGA III Exp. 21 Section 6.
+## Main definitions:
+ * `RootPairingCat`: Objects are root pairings.
+ * `Hom`: A morphism of root data is a linear map of weight spaces, its transverse on coweight
+   spaces, and a bijection on the set that indexes roots and coroots.
+ * `Hom.id`: The identity morphism.
+ * `Hom.comp`: The composite of two morphisms.
+
+## TODO
+
+ * Special types of morphisms: Isogenies, weight/coweight space embeddings
+ * Radicals
+ * Products and irreducibility
+ * Weyl group reimplementation?
+
+## Implementation details
+
+This is mostly copied from `ModuleCat`.
+
+-/
+
+open Set Function CategoryTheory
+
+noncomputable section
+
+universe v u
+
+variable {R : Type u} [CommRing R]
+
+/-- The category of root pairings. -/
+structure RootPairingCat (R : Type u) [CommRing R] where
+  /-- The weight space of a root pairing. -/
+  weight : Type v
+  [weightIsAddCommGroup : AddCommGroup weight]
+  [weightIsModule : Module R weight]
+  /-- The coweight space of a root pairing. -/
+  coweight : Type v
+  [coweightIsAddCommGroup : AddCommGroup coweight]
+  [coweightIsModule : Module R coweight]
+  /-- The set that indexes roots and coroots. -/
+  index : Type v
+  /-- The root pairing structure. -/
+  pairing : RootPairing index R weight coweight
+
+attribute [instance] RootPairingCat.weightIsAddCommGroup RootPairingCat.weightIsModule
+attribute [instance] RootPairingCat.coweightIsAddCommGroup RootPairingCat.coweightIsModule
+
+namespace RootPairingCat
+
+/-- A morphism of root pairings is a pair of mutually transposed maps of weight and coweight spaces
+that preserves roots and coroots.  We make the map of indexing sets explicit. -/
+@[ext]
+structure Hom (P Q : RootPairingCat R) where
+  /-- A linear map on weight space. -/
+  weight_map : P.weight →ₗ[R] Q.weight
+  /-- A contravariant linear map on coweight space. -/
+  coweight_map : Q.coweight →ₗ[R] P.coweight
+  /-- A bijection on index sets. -/
+  index_map : P.index ≃ Q.index
+  weight_coweight_transpose :
+    LinearMap.dualMap weight_map =
+      P.pairing.toDualRight ∘ₗ coweight_map ∘ₗ Q.pairing.toDualRight.symm
+  root_weight_map : weight_map ∘ P.pairing.root = Q.pairing.root ∘ index_map
+  coroot_coweight_map : coweight_map ∘ Q.pairing.coroot = P.pairing.coroot ∘ index_map.symm
```

Comment:
I think we should first define morphisms without using the category theory library.

For example we could create a new file (maybe called `RootSystem/Hom.lean`) and in it define:
```lean
structure RootPairing.Hom (ι₂ M₂ N₂ : Type*)
    [AddCommGroup M₂] [Module R M₂] [AddCommGroup N₂] [Module R N₂]
    (P : RootPairing ι R M N) (Q : RootPairing ι₂ R M₂ N₂) where
  /-- A linear map on weight space. -/
  weightMap : M →ₗ[R] M₂
  /-- A contravariant linear map on coweight space. -/
  coweightMap : N₂ →ₗ[R] N
  /-- A bijection on index sets. -/
  indexEquiv : ι ≃ ι₂
  weight_coweight_transpose : weightMap.dualMap ∘ₗ Q.toDualRight = P.toDualRight ∘ₗ coweightMap
  root_weightMap : weightMap ∘ P.root = Q.root ∘ indexEquiv
  coroot_coweightMap : coweightMap ∘ Q.coroot = P.coroot ∘ indexEquiv.symm
```
together with basic boilerplate and then use this to define the `Category` instance below (similar to what is done for the category of modules [here](https://github.com/leanprover-community/mathlib4/blob/00df099f8ed42c12b35cdedcfbf1fde6c7413662/Mathlib/Algebra/Category/ModuleCat/Basic.lean#L91-L97)).

I claim the advantages of this design are:

- We can have universe polymorphic morphisms
- It is possible to speak of morphisms without having to invoke the category theory library
- It is more idiomatic compared to the rest of our linear algebra

What do you think?

Reply:
No need IMHO, though I'm excited to see what comes next :)